src/lib.rs

```diff
@@ -0,0 +1,452 @@
+        let num_funcs;
+        let num_consts;
+        let num_stackmaps;
+        {
+            let mut cursor = cursor_from_elf(&elf_file, 0)?;
+            Self::check_header(&mut cursor)?;
+
+            // Read in table sizes.
+            // uint32: NumFunctions
+            num_funcs = cursor.read_u32::<NativeEndian>()?;
+            // uint32: NumConstants
+            num_consts = cursor.read_u32::<NativeEndian>()?;
+            // uint32: NumRecords
+            num_stackmaps = cursor.read_u32::<NativeEndian>()?;
+        }
+
+        Ok(Self{elf_file, num_funcs, num_consts, num_stackmaps})
+    }
+
+    /// Returns the number of stackmap record entries in the stackmap section.
+    pub fn num_stackmaps(&self) -> u32 {
+        self.num_stackmaps
+    }
+
+    /// Returns the number of function entries in the stackmap section.
+    pub fn num_funcs(&self) -> u32 {
+        self.num_funcs
+    }
+
+    /// Check the stackmap header looks sane.
+    fn check_header(cursor: &mut Cursor<&Vec<u8>>) -> SMParserResult<()> {
+        // uint8: Stack Map Version
+        let version = cursor.read_u8()?;
+        if version != STACKMAP_VERSION {
+            let msg = format!("Expected stackmap format v{} but binary is v{}", STACKMAP_VERSION, version);
+            return Err(SMParserError::Other(msg));
+        }
+        // uint8: Reserved (expected to be 0)
+        let b2 = cursor.read_u8()?;
+        if b2 != 0 {
+            let msg = format!("Expected 0 in stackmap section byte 2, got {}", b2);
+            return Err(SMParserError::Other(msg));
+        }
+        // uint16: Reserved (expected to be 0)
+        let b2_3 = cursor.read_u16::<NativeEndian>()?;
+        if b2_3 != 0 {
+            let msg = format!("Expected 0 in stackmap section bytes 2 and 3, got {}", b2_3);
+            return Err(SMParserError::Other(msg));
+        }
+        Ok(())
+    }
+
+    /// Make an iterator over the stackmap record entries in the stackmap section.
+    ///
+    /// If the iterator returns an error, the iterator becomes invalid.
+    ///
+    /// # Example
+    /// ```
+    /// use std::path::Path;
+    /// use yorick_smparser::StackMapParser;
+    ///
+    /// match StackMapParser::new(&Path::new("/bin/ls")) {
+    ///     // It's unlikey /bin/ls contains stackmaps, but you get the idea.
+    ///     Err(e) => println!("error: {}", e),
+    ///     Ok(p) =>  {
+    ///         for stmap_res in p.iter_stackmaps() {
+    ///             match stmap_res {
+    ///                 Ok(stmap) => println!("{:?}", stmap),
+    ///                 Err(e) => {
+    ///                     println!("error: {}", e);
+    ///                     break; // You must not re-use the iterator upon error.
+    ///                 }
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    pub fn iter_stackmaps(&self) -> SMRecIterator {
+        let start_pos = OFFS_STACK_SIZE_ENTRIES + u64::from(self.num_funcs) *
```

Comment:
This might be easier to read as `u64::from(self.num_funcs + SIZE_STACK_SIZE_ENTRY + ...)`.

Reply:
That will work as long as we do indeed promote `SIZE_STACK_SIZE_ENTRY` from `u8` to `u64`.

Most of this arithmetic constructions here are the suggestion of clippy, which asked me not to use `as u*` in fear of the semantics changing at a later date (!).

Reply:
To be clearer, you can't add things of different integer types without either an `as` cast or a `*::from` cast. So your suggestion works if all of the things inside are the same type.

Reply:
It's a suggestion: if it's not a good one, don't do it.

Reply:
I think we probably want to leave them as-is.